### PR TITLE
TASK: Improve Error Handling during Bulk Indexing

### PR DESCRIPTION
**Pull request by dfeyer**
Currently we log some errors but the CLI output don't warn the user about 
the bulk indexing errors. 

With this change applied an ErrorHandlingService is used by the node indexer 
and CLI controller to report and log internal errors.

### CODE REVIEW

**Comment by dfeyer** [comment]
This change also improve how we store the error message, before this change the raw error from Elastic is logged, now a more human readable report is build, like for Exception logging, check https://gist.github.com/dfeyer/6aff9df9ed7c7cf3346b46999d04b39b for an example, basically the report contains two sections Payload + Error, nicely formatted.

The CLI output look like this when bad things happens:
screenshot

**Comment by dfeyer** [comment]
To test the PR you can create an error in the groovy script, by example

File: Classes/Flowpack/ElasticSearch/ContentRepositoryAdaptor/Indexer/Error/BulkIndexingError.php [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+     * @var array
+     */
+    protected $errors;
+
+    /**
+     * @param array $currentBulkRequest
+     * @param array $errors
+     * @throws Exception
+     */
+    public function __construct(array $currentBulkRequest, $errors)
```
Comment (kdambekalns) [inline Classes/Flowpack/ElasticSearch/ContentRepositoryAdaptor/Indexer/Error/BulkIndexingError.php]: Should be `array $errors`

File: Classes/Flowpack/ElasticSearch/ContentRepositoryAdaptor/Indexer/Error/BulkIndexingError.php [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+
+        $referenceCode = date('YmdHis', $_SERVER['REQUEST_TIME']) . substr(md5(rand()), 0, 6);
+
+        $this->filename = FLOW_PATH_DATA . 'Logs/ElasticSearch/' . $referenceCode . '.txt';
+        $this->message = sprintf('Bulk indexing errors detected - See also: Data/Logs/ElasticSearch/%s on host: %s', basename($this->filename), gethostname());
+    }
+
+    /**
+     * Log the error message
+     */
```
Comment (kdambekalns) [inline Classes/Flowpack/ElasticSearch/ContentRepositoryAdaptor/Indexer/Error/BulkIndexingError.php]: Please add `@return void`

File: Classes/Flowpack/ElasticSearch/ContentRepositoryAdaptor/Indexer/Error/ErrorInterface.php [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+use TYPO3\Flow\Annotations as Flow;
+
+/**
+ * Error Interface
+ */
+interface ErrorInterface
+{
+    /**
+     * Log the error message
+     */
```
Comment (kdambekalns) [inline Classes/Flowpack/ElasticSearch/ContentRepositoryAdaptor/Indexer/Error/ErrorInterface.php]: Please add `@return void`

File: Classes/Flowpack/ElasticSearch/ContentRepositoryAdaptor/Indexer/Error/MalformedBulkRequestError.php [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+     */
+    public function __construct($message, array $tuple)
+    {
+        $this->message = $message;
+        $this->tuple = $tuple;
+    }
+
+    /**
+     * Log the error message
+     */
```
Comment (kdambekalns) [inline Classes/Flowpack/ElasticSearch/ContentRepositoryAdaptor/Indexer/Error/MalformedBulkRequestError.php]: Please add `@return void`

File: Classes/Flowpack/ElasticSearch/ContentRepositoryAdaptor/Service/ErrorHandlingService.php [hunk 5]
```diff
@@ -0,0 +1,50 @@
+<?php
+namespace Flowpack\ElasticSearch\ContentRepositoryAdaptor\Service;
+
+/*                                                                                                  *
+ * This script belongs to the TYPO3 Flow package "Flowpack.ElasticSearch.ContentRepositoryAdaptor". *
```
Comment (kdambekalns) [inline Classes/Flowpack/ElasticSearch/ContentRepositoryAdaptor/Service/ErrorHandlingService.php]: Whoah, that's an outdated file header… ;)

File: Classes/Flowpack/ElasticSearch/ContentRepositoryAdaptor/Service/ErrorHandlingService.php [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+use TYPO3\Flow\Annotations as Flow;
+
+/**
+ * Error Handling Service
+ *
+ * @Flow\Scope("singleton")
+ */
+class ErrorHandlingService implements \IteratorAggregate
+{
+    protected $errors = [];
```
Comment (kdambekalns) [inline Classes/Flowpack/ElasticSearch/ContentRepositoryAdaptor/Service/ErrorHandlingService.php]: Please add docblock…

File: Classes/Flowpack/ElasticSearch/ContentRepositoryAdaptor/Service/ErrorHandlingService.php [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+
+    /**
+     * @param ErrorInterface $error
```
Comment (kdambekalns) [inline Classes/Flowpack/ElasticSearch/ContentRepositoryAdaptor/Service/ErrorHandlingService.php]: Please add `@return void`

**Comment by dfeyer** [comment]
@kdambekalns The last commit should fix your comments, can you review again ? Nice if we can have this one merge for the release on the next major

**Comment by kdambekalns** [comment]
@dfeyer I have a merge of master into this ready here, but cannot push. Could you allow maintainer edits or grant me access otherwise?

**Comment by kdambekalns** [comment]
Hey, @dfeyer I'd really like to help out :)

**Comment by dfeyer** [comment]
@kdambekalns Push is now open

**Comment by kdambekalns** [comment]
Just a tiny amount of stuff merged, 315 commits ahead… ;)

**Comment by dfeyer** [comment]
Thanks for the refresh looks good

**Pull request merged**